django/views/i18n.py

```diff
@@ -318,3 +331,120 @@ def json_catalog(request, domain='djangojs', packages=None):
         'plural': plural,
     }
     return http.JsonResponse(data)
+
+
+class JavascriptCatalog(View):
+    """
+    Returns the selected language catalog as a javascript library.
+
+    Receives the list of packages to check for translations in the
+    kwargs `packages` key either from an infodict or as a +-delimited
+    string from the request. Default is 'django.conf'.
+
+    Additionally you can override the gettext domain for this view,
+    but usually you don't want to do that, as JavaScript messages
+    go to the djangojs domain. But this might be needed if you
+    deliver your JavaScript source from Django templates.
+    """
+    domain = 'djangojs'
+    packages = None
+
+    def get(self, request, *args, **kwargs):
+        locale = get_language()
+        domain = kwargs.get('domain', self.domain)
+        # If packages are not provided, default to all installed packages, as
+        # DjangoTranslation without localedirs is harvesting them all.
+        packages = kwargs.get('packages', '').split('+') or self.packages
+        paths = self.get_paths(packages) if packages else None
+        self.translation = DjangoTranslation(locale, domain=domain, localedirs=paths)
+        context = self.get_context_data(**kwargs)
+        return self.render_to_response(context)
+
+    def get_paths(self, packages):
+        app_configs = apps.get_app_configs()
+        allowable_packages = set(app_config.name for app_config in app_configs)
```

Comment:
Oh yes the tests! You are right then.






De: Claude Paroz
Enviado el: ‎miércoles‎, ‎09‎ de ‎marzo‎ de ‎2016 ‎19‎:‎23
Para: django/django
CC: Cristiano Coelho





In django/views/i18n.py:
> +    packages = None
> +
> +    def get(self, request, *args, **kwargs):
> +        locale = get_language()
> +        domain = kwargs.get('domain', self.domain)
> +        # If packages are not provided, default to all installed packages, as
> +        # DjangoTranslation without localedirs is harvesting them all.
> +        packages = kwargs.get('packages', '').split('+') or self.packages
> +        paths = self.get_paths(packages) if packages else None
> +        self.translation = DjangoTranslation(locale, domain=domain, localedirs=paths)
> +        context = self.get_context_data(**kwargs)
> +        return self.render_to_response(context)
> +
> +    def get_paths(self, packages):
> +        app_configs = apps.get_app_configs()
> +        allowable_packages = set(app_config.name for app_config in app_configs)


The tests could dynamically change installed apps, so I'm not sure the optimization is worth it, as I don't think this is a costly operation.

—
Reply to this email directly or view it on GitHub.